Document  geometrical transformations in user guide (#4385)
scikit-image/scikit-image: doc/source/user_guide/geometrical_transform.rst

============================================
Geometrical transformations of images
============================================

Cropping, resizing and rescaling images
---------------------------------------

.. currentmodule:: skimage.transform

Images being NumPy arrays (as described in the :ref:`numpy` section), cropping
an image can be done with simple slicing operations. Below we crop a 100x100

> square corresponding to the top left corner of the astronaut image. Note that

Reviewer: ```suggestion
square corresponding to the top-left corner of the astronaut image. Note that
```
to be consistent with line 51 of `doc/examples/transform/plot_transform_types.py`

> this operation is done for all color channels (the color dimension is the last

Reviewer: ```suggestion
this operation is done for all color channels (the color dimension is the last,
```

third dimension):: 

   >>> from skimage import data
   >>> img = data.astronaut()
   >>> top_left = img[:100, :100]

> In order to change the shape of the image,  :mod:`skimage.color` provides several

Reviewer: ```suggestion
In order to change the shape of the image, :mod:`skimage.color` provides several
```

functions described in :ref:`sphx_glr_auto_examples_transform_plot_rescale.py`
.

.. literalinclude:: ../../examples/transform/plot_rescale.py
    :language: python
    :start-after: import matplotlib.pyplot as plt
    :end-before: fig, axes

.. image:: ../auto_examples/transform/images/sphx_glr_plot_rescale_001.png
   :target: ../auto_examples/transform/plot_rescale.html
   :align: center

>    :width: 80%

Reviewer 2: I'm lost over here. The image I see is a 4x4 grayscale of the full version of the astronaut. It's not cropped, and not in color. Is this intended?

Author: This example shows how to use the functions `rescale`, `resize` and `downscale_local_mean`. It is not for cropping but rather changing the total size of the image. Do you think I should change the wording "n order to change the shape of the image"?

Projective transforms (homographies)
------------------------------------

`Homographies <https://en.wikipedia.org/wiki/Homography>`_
are transformations of a Euclidean space that preserve the alignment of points.
Specific cases of homographies correspond to the conservation of more properties,
such as parallelism (affine transformation), shape (similar transformation) or distances (Euclidean transformation).
The different types of homographies available in scikit-image are presented in 
:ref:`sphx_glr_auto_examples_transform_plot_transform_types.py`.

Projective transformations can either be created using the explicit
parameters (e.g. scale, shear, rotation and translation) or the full

> transformation matrix::

Reviewer 3: If you mention the transformation matrix, you should add example code with the matrix, as well as mention that it uses homogeneous coordinates.

Author: done, are you ok with the result?

   
   tform = tf.EuclideanTransform(
      rotation=np.pi / 12.,
      translation = (100, -20)
      )

Transformations can be applied to images using :func:`skimage.transform.warp`::

   img = img_as_float(data.chelsea())
   tf_img = tf.warp(img, tform.inverse)

.. image:: ../auto_examples/transform/images/sphx_glr_plot_transform_types_001.png
   :target: ../auto_examples/transform/plot_transform_types.html
   :align: center
   :width: 80%

The different transformations in :mod:`skimage.transform` have a ``estimate`` method in
order to estimate the parameters of the transformation from two sets of points
(the source and the destination), as explained in the :ref:`sphx_glr_auto_examples_transform_plot_geometric.py`
tutorial::

   text = data.text()

   src = np.array([[0, 0], [0, 50], [300, 50], [300, 0]])
   dst = np.array([[155, 15], [65, 40], [260, 130], [360, 95]])

   tform3 = tf.ProjectiveTransform()
   tform3.estimate(src, dst)
   warped = tf.warp(text, tform3, output_shape=(50, 300))

.. image:: ../auto_examples/transform/images/sphx_glr_plot_geometric_002.png
   :target: ../auto_examples/transform/plot_geometric.html
   :align: center
   :width: 80%

> The ``estimate`` method uses least-square optimization to minimize the distance

Reviewer 3: ```suggestion
The ``estimate`` method uses least-squares optimization to minimize the distance
```

between source and optimization.
Source and destination points can be determined manually, or using the
different methods for feature detection available in :mod:`skimage.feature`, 
such as

 * :ref:`sphx_glr_auto_examples_features_detection_plot_corner.py`,  
 * :ref:`sphx_glr_auto_examples_features_detection_plot_orb.py`,
 * :ref:`sphx_glr_auto_examples_features_detection_plot_brief.py`,
 * etc.

and matching points using :func:`skimage.feature.match_descriptors` before
estimating transformation parameters. However, spurious matches are often made, 

> and it is possible to use the RANSAC algorithm (instead of simple least-square optimization)

Reviewer 3: ```suggestion
and it is advisable to use the RANSAC algorithm (instead of simple least-squares optimization)
```

The different types of homographies available in scikit-image are presented in 
:ref:`sphx_glr_auto_examples_transform_plot_transform_types.py`.

Projective transformations can either be created using the explicit
parameters (e.g. scale, shear, rotation and translation)::

   from skimage import data
   from skimage import transform as tf
   from skimage import img_as_float
   
   tform = tf.EuclideanTransform(
      rotation=np.pi / 12.,
      translation = (100, -20)
      )

or the full transformation matrix::

   from skimage import data
   from skimage import transform as tf
   from skimage import img_as_float
   
   matrix = np.array([[np.cos(np.pi/12), -np.sin(np.pi/12), 100],
                      [np.sin(np.pi/12), np.cos(np.pi/12), -20],
                      [0, 0, 1]])
   tform = tf.EuclideanTransform(matrix)

The transformation matrix of a transform is available as its ``tform.params``
attribute. Transformations can be composed by multiplying matrices with the
``@`` matrix multiplication operator.

Transformation matrices use `Homogeneous coordinates <https://en.wikipedia.org/wiki/Homogeneous_coordinates>`_, which are the extension of Cartesian

> coordinates used in Euclidean geometry to the more general projective geometry. In particular, points at infinity can be represented with finite coordinates.

Reviewer 3: As above, wrap to 80c
